Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +370,124 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
-}
+    if (!has_Dirichlet) return;
+
+    // loop over all mesh refinement levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr   = phi[lev]->array(mfi);
+            // Extract tileboxes for which to loop
+            const Box& tb  = mfi.tilebox( phi[lev]->ixType().toIntVect() );
+
+            const auto lo = lbound(phi_arr);
+            const auto hi = ubound(phi_arr);
+
+            amrex::ParallelFor( tb,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    int idx, idx_lo, idx_hi;
+
+                    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+                        // check if the boundary in this dimension should be set
+                        if (!dirichlet_flag[idim]) return;
```

Comment:
Shouldn't this be `continue` rather than `return`? It seems like any dimension with no BC set will skip evaluation of the rest

Reply:
I believe the `amrex::ParallelFor` framework functions as a function call on each step rather than a conventional `for` loop which is why `return` is used rather than `continue`, but @ax3l could clarify that better.

Reply:
It'll act as a parallel loop over the parameters to the lambda, but not for loops inside, unless there's some really deep magic that I'm missing.

Reply:
Oh I see what you are saying now. I was lucky in that I always set the 0th dimension to be fixed so this didn't cause any problems. I'll change this now. Thanks for catching that!